tt_metal/api/event.hpp

```diff
@@ -0,0 +1,58 @@
+// SPDX-FileCopyrightText: © 2024 Tenstorrent Inc.
+//
+// SPDX-License-Identifier: Apache-2.0
+
+#pragma once
+
+#include "types.hpp"
+
+//==================================================
+//                  EVENT MANAGEMENT
+//==================================================
+
+namespace tt::tt_metal{
+namespace v1 {
+
+/**
+ * @brief Enqueues a command to record an event on the device.
+ *
+ * @param cq The command queue used to dispatch the command.
+ * @param event Shared pointer to the Event object to record.
+ */
+void EnqueueRecordEvent(CommandQueue cq, const std::shared_ptr<Event> &event);
+
+/**
+ * @brief Enqueues a command to wait for an event to complete on the device.
+ *
+ * @param cq The command queue that will wait for the event.
+ * @param event Shared pointer to the Event object to wait on.
+ */
+void EnqueueWaitForEvent(CommandQueue cq, const std::shared_ptr<Event> &event);
+
+/**
+ * @brief Blocks the host until the specified event has completed on the device.
+ *
+ * @param event Shared pointer to the Event object to synchronize.
+ */
+void EventSynchronize(const std::shared_ptr<Event> &event);
+
+/**
+ * @brief Queries the completion status of an event on the device.
+ *
+ * @param event Shared pointer to the Event object to query.
+ * @return True if the event is completed; otherwise, false.
+ */
+bool EventQuery(const std::shared_ptr<Event> &event);
+
```

Comment:
Make sure you want `const shared_ptr<T>` instead of `shared_ptr<const T>`. The former is an immutable pointer pointing to a mutable object. I don't think it is what you want after reading existing code (eventually casting to a `const Event&`). However, this is a common C++ API design issue and it's so common that people just don't/can't correct it.

Reply:
I don't think the intent here is to enforce immutability of `Event`, but rather to indicate the `std::shared_ptr`'s reference count will not be affected.